include/cuco/detail/static_set/kernels.cuh

```diff
@@ -245,6 +245,66 @@ __global__ void contains(InputIt first, cuco::detail::index_type n, OutputIt out
   }
 }
 
+/**
+ * @brief Finds the equivalent set elements of all keys in the range `[first, last)`.
+ *
+ * If the key `*(first + i)` has a match in the set, copies its matched element to `(output_begin +
+ * i)`. Else, copies the empty value sentinel. Uses the CUDA Cooperative Groups API to leverage
+ * groups of multiple threads to find each key. This provides a significant boost in throughput
+ * compared to the non Cooperative Group `find` at moderate to high load factors.
+ *
+ * @tparam CGSize Number of threads in each CG
+ * @tparam BlockSize The size of the thread block
+ * @tparam InputIt Device accessible input iterator
+ * @tparam OutputIt Device accessible output iterator assignable from the set's `value_type`
+ * @tparam Ref Type of non-owning device ref allowing access to storage
+ *
+ *
+ * @param first Beginning of the sequence of keys
+ * @param n Number of keys to query
+ * @param output_begin Beginning of the sequence of matched elements retrieved for each key
+ * @param ref Non-owning set device ref used to access the slot storage
+ */
+template <int32_t CGSize, int32_t BlockSize, typename InputIt, typename OutputIt, typename Ref>
+__global__ void find(InputIt first, cuco::detail::index_type n, OutputIt output_begin, Ref ref)
+{
+  namespace cg = cooperative_groups;
+
+  auto const block      = cg::this_thread_block();
+  auto const thread_idx = block.thread_rank();
+
+  cuco::detail::index_type const loop_stride = gridDim.x * BlockSize / CGSize;
+  cuco::detail::index_type idx               = (BlockSize * blockIdx.x + threadIdx.x) / CGSize;
+  __shared__ typename Ref::value_type output_buffer[BlockSize / CGSize];
+
+  while (idx - thread_idx < n) {  // the whole thread block falls into the same iteration
+    if (idx < n) {
+      auto const key = *(first + idx);
+      if constexpr (CGSize == 1) {
+        auto const found = ref.find(key);
+        /*
+         * The ld.relaxed.gpu instruction used in ref.find causes L1 to
+         * flush more frequently, causing increased sector stores from L2 to global memory.
+         * By writing results to shared memory and then synchronizing before writing back
+         * to global, we no longer rely on L1, preventing the increase in sector stores from
+         * L2 to global and improving performance.
+         */
```

Comment:
Is this still relevant? I test this in the CG code path and see a slight boost without sync + the use of shared memory buffer (see the code below).

Reply:
Sorry about the confusion, the optimization here is to use a shared memory buffer and each thread writes into the buffer first, then does a block-wise synchronization and in the end writes into the output. I noticed this effort is only valid for non-CG algorithms and for the CG code path, directly writing into the output without using synchronization or shared memory buffer can consistently save about 0.5% of runtime.